File: 프로그래머스/3/86053. 금과 은 운반하기/금과 은 운반하기.py
def solution(a, b, g, s, w, t):
    [low, high] = [0, 10**15]
    answer = high
    
    while low <= high:
        mid = (low + high) // 2
        if chech(mid, a, b, g, s, w, t):
            answer = mid
            high = mid - 1
        else:
            low = mid + 1
    
    return answer

def chech(time, a, b, g, s, w, t):
    [total_g, total_s, total_total] = [0] * 3
    
    for i in range(len(g)):
        round_time = 2 * t[i]
        max_count = (time + t[i]) // round_time
        max_weight = max_count * w[i]
        
        total_g += min(g[i], max_weight)
        total_s += min(s[i], max_weight)
        total_total += min(g[i] + s[i], max_weight)
    
    return total_g >= a and total_s >= b and total_total >= (a + b)

File: 프로그래머스/3/86053. 금과 은 운반하기/금과 은 운반하기.js
function solution(a, b, g, s, w, t) {
    let low = 0;
    let high = 10**14 * 4; 
    let answer = high;

    while (low <= high) {
        let mid = Math.floor((low + high) / 2);
        
        if (check(mid, a, b, g, s, w, t)) {
            answer = mid;
            high = mid - 1; // 가능하다면 시간을 더 줄인다.
        } else {
            low = mid + 1; // 불가능하다면 시간을 더 늘린다.
        }
    }

    return answer;
}

function check(time, a, b, g, s, w, t) {
    let total_g = 0;
    let total_s = 0;
    let total_total = 0;
    
    for (let i = 0; i < g.length; i++) {
        let roundTime = t[i] * 2;
        let moveCount = Math.floor((time + t[i]) / roundTime);
        
        let maxWeight = moveCount * w[i];

        total_g += Math.min(g[i], maxWeight);
        total_s += Math.min(s[i], maxWeight);
        total_total += Math.min(g[i] + s[i], maxWeight);
    }

    // 해당 시간 내에 나를 수 있는 금의 총합 ≥ 날라야 하는 금의 양
    // 해당 시간 내에 나를 수 있는 은의 총합 ≥ 날라야 하는 은의 양
    // 해당 시간 내에 나를 수 있는 (금+은) 총합 ≥ 날라야 하는 (금+은)의 양
    if (total_g >= a && total_s >= b && total_total >= a + b) {
        return true;
    }
    return false;
}